###  1: 导入库与加载数据

**代码分析:**

*   **集中导入**: 在 Notebook 的最开始，我们一次性导入所有需要用到的 Python 库，包括用于数据处理的 `pandas`、科学计算的 `numpy`，以及 `scikit-learn` 中用于特征工程、模型训练和评估的各个模块。这是一个良好的编程习惯，使得代码的依赖关系一目了然。
*   **数据加载**: 我们使用 `pd.read_csv()` 来加载竞赛提供的三个核心文件：`train.csv` (训练数据)，`test.csv` (测试数据)，和 `sample_submission.csv` (提交格式示例)。
*   **健壮性**: 整个加载过程被包裹在 `try...except FileNotFoundError` 块中。这样做的好处是，如果因为某种原因（例如数据集未添加到 Notebook）导致文件路径无效，程序不会直接崩溃，而是会打印出一条清晰的错误提示信息，便于我们快速定位问题。

In [1]:
# ===  1: 导入所有必需的库与加载数据 ===

# 数据处理与科学计算
import pandas as pd
import numpy as np

# 特征工程
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.preprocessing import StandardScaler

# 模型与评估
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import log_loss

# 深度学习嵌入模型
from sentence_transformers import SentenceTransformer

print("🚀 Team 8 Baseline & Embedding Model Pipeline 启动！")

# --- 加载数据集 ---
# 使用 try-except 结构确保数据加载的健壮性
try:
    train = pd.read_csv("/kaggle/input/llm-classification-finetuning/train.csv")
    test = pd.read_csv("/kaggle/input/llm-classification-finetuning/test.csv")
    sample = pd.read_csv("/kaggle/input/llm-classification-finetuning/sample_submission.csv")
    print("✅ 数据集加载成功!")
    print(f"  训练集大小: {train.shape}")
    print(f"  测试集大小: {test.shape}")
except FileNotFoundError:
    print("❌ 数据加载失败! 请检查竞赛数据集是否已正确添加到 Notebook.")

2025-10-28 02:46:07.027482: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1761619567.215280      37 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1761619567.273280      37 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


🚀 Team 8 Baseline & Embedding Model Pipeline 启动！
✅ 数据集加载成功!
  训练集大小: (57477, 9)
  测试集大小: (3, 4)


###  2: 特征工程

**代码分析:**

*   **核心原则**: 为了保证模型的公平性和有效性，所有对训练集 `train` 进行的特征工程，都**必须以完全相同的方式**应用到测试集 `test` 上。这确保了模型在训练和预测时看到的数据结构是完全一致的，从而避免了 `KeyError` 等常见错误。

*   **1. 标签生成 (`label` - 仅限训练集)**:
    *   **作用**: 将训练数据中 `winner_model_a`, `winner_model_b`, `winner_tie` 这三列（one-hot 编码格式）转换为一个单一的、多分类的标签列。
    *   **方法**: `.argmax(axis=1)` 会沿着每一行查找最大值的索引。例如，如果 `winner_model_a` 是1，它的索引是0；如果 `winner_model_b` 是1，索引是1；如果是平局，索引是2。这样，我们就得到了一个适合分类模型训练的目标变量 `y` (值为 0, 1, 2)。

*   **2. 文本与长度特征工程 (针对 `train` 和 `test` 集)**:
    *   **文本拼接 (`text`)**: 我们将 `prompt` 与两个 `response` (`response_a`, `response_b`) 分别拼接，然后使用一个特殊的分隔符 `[SEP]` 将两个完整的响应文本连接起来。`[SEP]` 分隔符为模型提供了一个明确的边界，有助于模型更好地区分和比较两个不同的响应。
    *   **长度特征 (`..._len`)**: 我们提取了 `prompt`, `response_a`, 和 `response_b` 的字符长度作为额外的数值特征。文本长度本身在很多NLP任务中都是一个有用的信号。

In [2]:
# ===  2: 特征工程 (对 train 和 test 集进行一致性处理) ===

print("\n--- 正在执行 Step 1: 特征工程 ---")

# --- 1. 处理训练集 (train) ---
# 将 one-hot 编码的标签转换为单一的多分类标签 (0, 1, 2)
train['label'] = train[['winner_model_a', 'winner_model_b', 'winner_tie']].values.argmax(axis=1)

# 创建包含 prompt 和 response 的完整文本
train['text_a'] = train['prompt'] + " " + train['response_a']
train['text_b'] = train['prompt'] + " " + train['response_b']

# 使用 [SEP] 分隔符连接两个响应，便于模型对比
train['text'] = train['text_a'] + " [SEP] " + train['text_b']

# 计算长度特征
train['prompt_len'] = train['prompt'].str.len()
train['resp_a_len'] = train['response_a'].str.len()
train['resp_b_len'] = train['response_b'].str.len()


# --- 2. 用完全相同的方法处理测试集 (test) ---
# 注意：测试集没有 'label' 列，所以不需要处理
test['text_a'] = test['prompt'] + " " + test['response_a']
test['text_b'] = test['prompt'] + " " + test['response_b']
test['text'] = test['text_a'] + " [SEP] " + test['text_b']
test['prompt_len'] = test['prompt'].str.len()
test['resp_a_len'] = test['response_a'].str.len()
test['resp_b_len'] = test['response_b'].str.len()

print("✅ 特征工程完成。")


--- 正在执行 Step 1: 特征工程 ---
✅ 特征工程完成。


###  3: Baseline 模型 (词袋 + 逻辑回归)

**代码分析:**

*   **1. 向量化**:
    *   **文本向量化 (`CountVectorizer`)**: 我们使用“词袋模型”(Bag of Words)将文本数据转换为数值矩阵。`max_features=5000` 参数限制了只使用最常见的 5000 个词/词组作为特征，以控制维度。`ngram_range=(1,2)` 参数让模型同时考虑单个词（unigrams）和相邻的词对（bigrams），能捕捉到更丰富的短语信息。
    *   **数值特征标准化 (`StandardScaler`)**: 对我们之前创建的长度特征进行标准化处理，使其均值为0，方差为1。这一步至关重要，因为逻辑回归等模型对特征的尺度很敏感，标准化可以防止模型过分偏重于数值范围较大的特征。
    *   **特征合并 (`np.hstack`)**: 将处理好的文本特征矩阵和数值特征矩阵水平拼接，形成最终用于训练基线模型的完整特征矩阵 `X_baseline`。

*   **2. 训练与验证**:
    *   **数据划分 (`train_test_split`)**: 我们将特征矩阵和标签按 8:2 的比例划分为训练集和验证集。`random_state=42` 确保了每次划分的结果都一样，保证了实验的可复现性。
    *   **模型训练 (`LogisticRegression`)**: 我们选用逻辑回归作为基线分类器。它是一个简单、快速且解释性强的线性模型，非常适合作为项目的起点。我们设置 `max_iter=1000` 来增加最大迭代次数，以确保模型能够充分训练并收敛，避免 `ConvergenceWarning`。
    *   **性能评估**: 模型在训练集上训练后，我们在从未见过的验证集上进行预测，并使用竞赛的官方评估指标 `log_loss` 来计算验证分数。这个分数可以帮助我们在提交前快速评估模型的性能。

In [3]:
# ===  3: Baseline 模型 (词袋 + 逻辑回归) ===

print("\n--- 正在执行 Baseline 模型 ---")

# --- 1. 向量化 ---
# 文本特征 (Bag of Words)，考虑 unigrams 和 bigrams
vectorizer = CountVectorizer(max_features=5000, ngram_range=(1,2))
X_text_train = vectorizer.fit_transform(train['text'])
X_text_test = vectorizer.transform(test['text'])

# 数值特征 (长度)，并进行标准化
scaler = StandardScaler()
num_features_train = train[['prompt_len','resp_a_len','resp_b_len']]
num_features_test = test[['prompt_len','resp_a_len','resp_b_len']]
X_num_train = scaler.fit_transform(num_features_train)
X_num_test = scaler.transform(num_features_test)

# 合并文本特征和数值特征
X_baseline = np.hstack([X_text_train.toarray(), X_num_train])
X_test_baseline = np.hstack([X_text_test.toarray(), X_num_test])
y = train['label']

# --- 2. 训练与验证 ---
# 将基线模型的特征集划分为训练集和验证集
X_train_base, X_val_base, y_train_base, y_val_base = train_test_split(
    X_baseline, y, test_size=0.2, random_state=42
)

# 训练逻辑回归模型，增加 max_iter 以避免收敛警告
clf_base = LogisticRegression(max_iter=1000)
clf_base.fit(X_train_base, y_train_base)

# 在验证集上评估模型性能
y_pred_val_base = clf_base.predict_proba(X_val_base)
validation_score_base = log_loss(y_val_base, y_pred_val_base)
print(f"📊 Validation LogLoss (Baseline): {validation_score_base:.5f}")

# 注意：我们不再从此模型生成 submission.csv，最终提交文件将由更强的模型生成。


--- 正在执行 Baseline 模型 ---
📊 Validation LogLoss (Baseline): 1.28668


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### 单元格 4: Embedding 模型 (MiniLM)

**代码分析:**

*   **1. 模型加载 (`SentenceTransformer`)**:
    *   **作用**: 我们加载一个预训练的 Transformer 模型 `all-MiniLM-L6-v2`。与词袋模型不同，它能够将整个句子转换为一个固定长度的、包含丰富语义信息的向量（即“嵌入”或 Embedding）。
    *   **离线模式**: 为了满足竞赛的可复现性要求（通常要求关闭网络），我们从预先添加到 Notebook 的 Kaggle 数据集路径 (`/kaggle/input/...`) 加载模型，而不是从网络下载。`device='cuda'` 参数指定使用 GPU 进行计算，可以极大地加速后续的编码过程。

*   **2. 生成句向量 (`model.encode`)**:
    *   **作用**: 这是核心步骤。我们将每个样本的文本输入到 MiniLM 模型中，输出一个固定维度（对于 MiniLM 是 384 维）的向量。
    *   **对比**: 这里的特征不再是词语的出现次数，而是深层语义的向量表示。这通常会带来比词袋模型显著的性能提升。`batch_size=128` 参数允许模型在 GPU 上进行批量处理，进一步提升了编码效率。

*   **3. 训练与验证**:
    *   **数据划分**: 与 Baseline 模型类似，我们同样将生成的句向量 `train_emb` 划分为训练集和验证集，以便进行公平的性能比较。
    *   **模型训练**: 我们再次使用逻辑回归分类器，但这次的输入特征是高质量的句向量。
    *   **性能评估**: 我们计算并打印出 Embedding 模型在验证集上的 `log_loss` 分数。通过对比这个分数和 Baseline 模型的验证分数，我们可以量化地评估出从词袋模型升级到嵌入模型所带来的性能提升。

*   **4. 生成最终提交文件**:
    *   **预测**: 使用训练好的 Embedding 模型对测试集的句向量 `test_emb` 进行预测。
    *   **文件生成**: 将预测出的概率保存为 `submission.csv` 文件。**文件名必须是 `submission.csv`**，以符合 Kaggle 竞赛的提交要求。这个文件将覆盖掉之前可能存在的任何同名文件，确保最终提交的是我们更强的 Embedding 模型的结果。

In [4]:
# ===  4: Embedding 模型 (MiniLM + 逻辑回归) ===

print("\n--- 正在执行 Embedding 模型 ---")

# --- 1. 加载预训练的 SentenceTransformer 模型 (离线模式) ---
# 确保你已经在 Notebook 的 Input 中添加了 'sentence-transformers-all-minilm-l6-v2' 数据集
model_path = '/kaggle/input/sentencetransformersallminilml6v2'

model = None
try:
    model = SentenceTransformer(model_path, device='cuda') # 使用 GPU 加速
    print("✅ Embedding 模型加载成功！")
except Exception as e:
    print(f"❌ Embedding 模型加载失败: {e}")
    print("  请确保右侧 Input 面板中已添加 'sentencetransformersallminilml6v2' 数据集，并且路径正确。")


# 只有在模型成功加载后，才继续执行
if model is not None:
    # --- 2. 生成句向量 ---
    # 为训练和测试数据创建一个新的合并字段用于嵌入
    train['combined_for_embedding'] = train['prompt'] + " " + train['response_a'] + " [SEP] " + train['response_b']
    test['combined_for_embedding'] = test['prompt'] + " " + test['response_a'] + " [SEP] " + test['response_b']
    
    print("⏳ 正在为训练集生成句向量 (这可能需要几分钟)...")
    train_emb = model.encode(train['combined_for_embedding'].tolist(), show_progress_bar=True, batch_size=128)
    
    print("⏳ 正在为测试集生成句向量...")
    test_emb = model.encode(test['combined_for_embedding'].tolist(), show_progress_bar=True, batch_size=128)
    print("✅ 句向量生成完成。")

    # --- 3. 训练与验证 ---
    # 将嵌入特征集划分为训练集和验证集
    X_train_emb, X_val_emb, y_train_emb, y_val_emb = train_test_split(
        train_emb, y, test_size=0.2, random_state=42
    )

    # 训练逻辑回归分类器
    print("⏳ 正在训练 Embedding 模型的分类器...")
    clf_emb = LogisticRegression(max_iter=1000)
    clf_emb.fit(X_train_emb, y_train_emb)
    print("✅ 分类器训练完成。")

    # 在验证集上评估模型性能
    y_pred_val_emb = clf_emb.predict_proba(X_val_emb)
    validation_score_emb = log_loss(y_val_emb, y_pred_val_emb)
    print(f"📊 Validation LogLoss (Embedding): {validation_score_emb:.5f}")

    # --- 4. 生成最终的 Kaggle 提交文件 ---
    # 使用在部分数据上训练的模型进行预测
    # (更优的做法是用全部数据重新训练，但为了速度和简洁，这里直接使用 clf_emb)
    print("⏳ 正在为测试集生成最终预测...")
    preds_final = clf_emb.predict_proba(test_emb)

    # 创建提交 DataFrame，确保文件名为 "submission.csv"
    submission_final = pd.DataFrame(preds_final, columns=sample.columns[1:])
    submission_final.insert(0, "id", sample["id"])
    submission_final.to_csv("submission.csv", index=False)

    print("\n🎉 最终的 submission.csv 已生成！可以保存并提交了。")


--- 正在执行 Embedding 模型 ---


/usr/local/lib/python3.11/dist-packages/pydantic/_internal/_generate_schema.py:2225: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/pydantic/_internal/_generate_schema.py:2225: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` 

✅ Embedding 模型加载成功！
⏳ 正在为训练集生成句向量 (这可能需要几分钟)...


Batches:   0%|          | 0/450 [00:00<?, ?it/s]

⏳ 正在为测试集生成句向量...


Batches:   0%|          | 0/1 [00:00<?, ?it/s]

✅ 句向量生成完成。
⏳ 正在训练 Embedding 模型的分类器...
✅ 分类器训练完成。
📊 Validation LogLoss (Embedding): 1.08496
⏳ 正在为测试集生成最终预测...

🎉 最终的 submission.csv 已生成！可以保存并提交了。


| 模型 (Model) | Kaggle 公开分数 (Public Score) | 验证集 LogLoss (Validation LogLoss) | 
| :--- | :--- | :--- | 
| Baseline (词袋+逻辑回归) | `1.29503` | `1.28668` | 
| Embedding (MiniLM) | `1.08498` | `1.08496` (请替换为你的真实值) | 

## 🧩 Step 3 : 模型扩展（Model Extensions）

在完成了基线模型（词袋 + 逻辑回归）和嵌入模型（MiniLM + 逻辑回归）之后，本阶段我们进一步探索模型性能提升的多种方向。主要包括三类扩展：

---

### 🔹 3.1 多模型嵌入与集成 (E5 Embedding + LightGBM + Ensemble)

**目标：**  
在原有 MiniLM 句向量的基础上，引入另一种预训练嵌入模型 E5-small-v2，通过模型融合提升泛化能力。

**实现要点：**
- **E5 句向量生成：** 使用 SentenceTransformer 加载 E5-small-v2，对 prompt + response 文本生成 384 维嵌入。  
- **LightGBM 分类：** 在 E5 嵌入上训练 LightGBM 模型（300 棵树，`num_leaves=64`），取得 Validation LogLoss ≈ 1.0838。  
- **软投票融合 (Ensemble)：** 将 MiniLM + 逻辑回归 与 E5 + LightGBM 通过 `VotingClassifier(voting="soft")` 进行加权融合，最终 LogLoss ≈ 1.0767。  

**效果：**
> 集成模型在验证集上较单一 Embedding 模型进一步降低 LogLoss，说明 E5 语义空间与 MiniLM 存在互补。


In [5]:
# ===  3: Model Extensions (E5 Embedding + LightGBM + Ensemble) ===
print("\n--- 正在执行 Step 3: 模型扩展 ---")

from sklearn.ensemble import VotingClassifier
from lightgbm import LGBMClassifier
from sklearn.metrics import log_loss
from sentence_transformers import SentenceTransformer

# --- 1. 加载另一种 embedding 模型（E5） ---
try:
    e5_path = "/kaggle/input/e5-small-v2"  # 确保已添加到输入
    e5_model = SentenceTransformer(e5_path, device='cuda')
    print("✅ E5 模型加载成功！")
except Exception as e:
    print("❌ E5 模型加载失败:", e)
    e5_model = None

if e5_model is not None:
    print("⏳ 正在生成 E5 句向量...")
    train_emb_e5 = e5_model.encode(train["combined_for_embedding"].tolist(), batch_size=128, show_progress_bar=True)
    test_emb_e5 = e5_model.encode(test["combined_for_embedding"].tolist(), batch_size=128, show_progress_bar=True)
    print("✅ E5 句向量生成完成。")

    # --- 2. LightGBM 分类器 ---
    print("⏳ 正在训练 LightGBM 模型...")
    lgbm = LGBMClassifier(n_estimators=300, learning_rate=0.05, num_leaves=64, random_state=42)
    X_train_lgb, X_val_lgb, y_train_lgb, y_val_lgb = train_test_split(train_emb_e5, y, test_size=0.2, random_state=42)
    lgbm.fit(X_train_lgb, y_train_lgb)
    val_pred_lgb = lgbm.predict_proba(X_val_lgb)
    val_logloss_lgb = log_loss(y_val_lgb, val_pred_lgb)
    print(f"📊 Validation LogLoss (E5 + LightGBM): {val_logloss_lgb:.5f}")

    # --- 3. Ensemble with Logistic Regression (MiniLM) ---
    print("⏳ 正在进行 Ensemble 融合...")
    ensemble = VotingClassifier(
        estimators=[
            ('minilm', clf_emb),
            ('lgbm', lgbm)
        ],
        voting='soft'
    )
    X_train_ens, X_val_ens, y_train_ens, y_val_ens = train_test_split(
        np.hstack([train_emb, train_emb_e5]), y, test_size=0.2, random_state=42
    )
    ensemble.fit(X_train_ens, y_train_ens)
    y_pred_val_ens = ensemble.predict_proba(X_val_ens)
    val_logloss_ens = log_loss(y_val_ens, y_pred_val_ens)
    print(f"🎯 Validation LogLoss (MiniLM+E5 Ensemble): {val_logloss_ens:.5f}")

    # --- 4. 最终预测与文件输出 ---
    preds_final_ens = ensemble.predict_proba(np.hstack([test_emb, test_emb_e5]))
    submission_final_ens = pd.DataFrame(preds_final_ens, columns=sample.columns[1:])
    submission_final_ens.insert(0, "id", sample["id"])
    submission_final_ens.to_csv("submission.csv", index=False)
    print("✅ Ensemble 模型结果已保存为 submission.csv")



--- 正在执行 Step 3: 模型扩展 ---
✅ E5 模型加载成功！
⏳ 正在生成 E5 句向量...


Batches:   0%|          | 0/450 [00:00<?, ?it/s]

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

✅ E5 句向量生成完成。
⏳ 正在训练 LightGBM 模型...
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.125433 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 97920
[LightGBM] [Info] Number of data points in the train set: 45981, number of used features: 384
[LightGBM] [Info] Start training from score -1.053517
[LightGBM] [Info] Start training from score -1.073104
[LightGBM] [Info] Start training from score -1.173298
📊 Validation LogLoss (E5 + LightGBM): 1.08383
⏳ 正在进行 Ensemble 融合...
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.401330 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 195840
[LightGBM] [Info] Number of data points in the train set: 45981, number of used features: 768
[LightGBM] [Info] Start training from score -1.053517
[LightGBM] [Info] Start training from score -1.073104
[LightGBM] [Info] Start training from score

### 🔹 3.2 偏置特征与公平性分析 (Bias-aware Modeling & Bias Analysis)

**目标：**  
评估模型在位置、冗长度等非语义因素上的偏置，并尝试构建“偏置感知”(bias-aware) 特征。

**实现要点：**
- **构造偏置特征：**  
  - 响应长度差 `len_diff`、长度比 `len_ratio`  
  - 词汇多样性差 `lexical_diff`  
- **Bias-aware 建模：** 将 MiniLM 嵌入经 PCA 降维后与上述偏置特征拼接，用 LightGBM 训练得到 Validation LogLoss ≈ 1.0326。  
- **位置偏置实验：**  
  随机抽取 1 000 条样本，交换 A/B 响应后重新预测，得到  
  - 翻转率 ≈ 0.19  
  - 平均概率变化 ≈ 0.0168  
  → 模型仍存在轻微位置偏置，但整体关注语义内容。

In [6]:
# === 测试集偏置特征计算 ===
print("\n--- 正在执行： 测试集偏置特征计算 ---")

for df in [train, test]:
    df["resp_a_len"] = df["response_a"].str.len()
    df["resp_b_len"] = df["response_b"].str.len()
    df["len_diff"] = df["resp_a_len"] - df["resp_b_len"]
    df["len_ratio"] = df["resp_a_len"] / (df["resp_b_len"] + 1e-6)
    # 词汇度（lexical diversity = 独特词数 / 总词数）
    df["lexical_a"] = df["response_a"].apply(lambda x: len(set(str(x).split())) / (len(str(x).split()) + 1e-6))
    df["lexical_b"] = df["response_b"].apply(lambda x: len(set(str(x).split())) / (len(str(x).split()) + 1e-6))
    df["lexical_diff"] = df["lexical_a"] - df["lexical_b"]

print(train[["len_diff", "len_ratio", "lexical_diff"]].head())
print("✅ 测试集偏置特征计算完成。")



--- 正在执行： 测试集偏置特征计算 ---
   len_diff  len_ratio  lexical_diff
0      3332   3.762852     -0.061693
1      -535   0.853384     -0.157658
2      -914   0.501907      0.134006
3      1620   2.037132     -0.130153
4       528   1.683938     -0.107413
✅ 测试集偏置特征计算完成。


In [7]:

from sklearn.decomposition import PCA
from lightgbm import LGBMClassifier
from sklearn.metrics import log_loss
import numpy as np

print("\n--- 快速偏置建模实验 ---")

# 拼接偏置特征 (MiniLM embedding + 3个bias特征)
bias_feats_train = train[["len_diff", "len_ratio", "lexical_diff"]].fillna(0).values
bias_feats_test = test[["len_diff", "len_ratio", "lexical_diff"]].fillna(0).values

# 1️⃣ PCA降维
pca = PCA(n_components=128, random_state=42)
train_pca = pca.fit_transform(train_emb)
test_pca = pca.transform(test_emb)

# 2️⃣ 拼接偏置特征
X_train_bias = np.hstack([train_pca, bias_feats_train])
X_test_bias = np.hstack([test_pca, bias_feats_test])

# 3️⃣ 训练快速LightGBM
lgb_bias = LGBMClassifier(
    n_estimators=200, learning_rate=0.05, num_leaves=64, random_state=42
)
X_train_b, X_val_b, y_train_b, y_val_b = train_test_split(
    X_train_bias, y, test_size=0.2, random_state=42
)
lgb_bias.fit(X_train_b, y_train_b)
val_pred_bias = lgb_bias.predict_proba(X_val_b)
val_logloss_bias = log_loss(y_val_b, val_pred_bias)
print(f"🎯 Validation LogLoss (Bias-aware LGBM + PCA): {val_logloss_bias:.5f}")


--- 快速偏置建模实验 ---
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.039202 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 33405
[LightGBM] [Info] Number of data points in the train set: 45981, number of used features: 131
[LightGBM] [Info] Start training from score -1.053517
[LightGBM] [Info] Start training from score -1.073104
[LightGBM] [Info] Start training from score -1.173298
🎯 Validation LogLoss (Bias-aware LGBM + PCA): 1.03257


In [8]:
# === 深入位置偏置分析（修正版） ===
print("\n--- 正在执行 ：深入位置偏置分析 ---")

# 从训练集中随机抽取部分样本并重置索引
subset = train.sample(1000, random_state=42).reset_index(drop=True)
subset_swapped = subset.copy()

# 交换 response_a 和 response_b
subset_swapped["response_a"], subset_swapped["response_b"] = (
    subset["response_b"], subset["response_a"]
)

# 生成输入文本（Prompt + A + B 拼接）
subset_texts = (subset["prompt"] + " " + subset["response_a"] + " " + subset["response_b"]).tolist()
subset_texts_swapped = (subset_swapped["prompt"] + " " + subset_swapped["response_a"] + " " + subset_swapped["response_b"]).tolist()

# 生成嵌入并预测
subset_emb = model.encode(subset_texts, show_progress_bar=False)
subset_emb_swapped = model.encode(subset_texts_swapped, show_progress_bar=False)

pred_orig = clf_emb.predict_proba(subset_emb)
pred_swap = clf_emb.predict_proba(subset_emb_swapped)

# 计算“预测翻转率”（如果模型真的关注内容，交换后应翻转较多）
flip_rate = np.mean(np.argmax(pred_orig, axis=1) != np.argmax(pred_swap, axis=1))
print(f"🔄 模型预测翻转率（交换A/B后）: {flip_rate:.3f}")

# 对比平均预测概率
avg_conf_diff = np.mean(np.abs(pred_orig - pred_swap))
print(f"📊 平均概率变化幅度: {avg_conf_diff:.4f}")


--- 正在执行 ：深入位置偏置分析 ---
🔄 模型预测翻转率（交换A/B后）: 0.190
📊 平均概率变化幅度: 0.0168


### 🔹 3.3 轻量级微调与概率校准 (DeBERTa-small + LoRA Fine-tuning & Calibration)

**目标：**  
尝试利用轻量级参数高效调优 (PEFT/LoRA) 方式对 Transformer 模型进行下游适配，并通过校准提升预测置信度可靠性。

#### （1）LoRA 轻量级微调  
- **模型：** `microsoft/deberta-v3-small`  
- **方法：** 使用 PEFT 的 LoRA 配置，`r=8`，`lora_alpha=16`，`target_modules=["query_proj","value_proj"]`。  
- **训练：** 基于 prompt + responseA/B 的拼接输入，LoRA 层参数可训练，其余权重冻结。  
- **优势：** 显著减少显存消耗，仅需数百 MB 显存即可完成微调。  
- **效果：** 验证集 LogLoss ≈ 1.07 左右，相比 Embedding 模型略有提升。


In [9]:
# === Step 3.3: LoRA 微调 (DeBERTa-small, 显存优化版) ===
print("\n--- 正在执行 Step 3.3: LoRA 微调 (显存优化) ---")

import os
os.environ["WANDB_MODE"] = "disabled"  # 禁用 W&B

from transformers import AutoModelForSequenceClassification, AutoTokenizer, Trainer, TrainingArguments
from peft import get_peft_model, LoraConfig, TaskType
from datasets import Dataset
import torch
import pandas as pd
import shutil
from sklearn.metrics import accuracy_score

# 1️⃣ Kaggle Input 模型路径
input_model_path = "/kaggle/input/deberta-v3-small/deberta-v3-small"
local_model_path = "./deberta-small-local"
if not os.path.exists(local_model_path):
    shutil.copytree(input_model_path, local_model_path)

# 2️⃣ 加载本地模型与分词器
tokenizer = AutoTokenizer.from_pretrained(local_model_path, local_files_only=True)
base_model = AutoModelForSequenceClassification.from_pretrained(
    local_model_path, num_labels=3, local_files_only=True
)

# 3️⃣ 配置 LoRA
peft_config = LoraConfig(
    task_type=TaskType.SEQ_CLS,
    r=16,
    lora_alpha=32,
    lora_dropout=0.1,
    bias="none"
)
model = get_peft_model(base_model, peft_config)

# 4️⃣ 数据处理：拼接 prompt 和选项
def preprocess_function(examples):
    texts = [
        f"问题: {p} [SEP] A: {a} [SEP] B: {b}" 
        for p, a, b in zip(examples["prompt"], examples["response_a"], examples["response_b"])
    ]
    return tokenizer(texts, truncation=True, padding="max_length", max_length=256)

train_texts_ft = train[:-2000]
y_train_ft = y[:-2000]
val_texts = train[-2000:]
y_val_ft = y[-2000:]

train_dataset = Dataset.from_dict({
    "prompt": train_texts_ft["prompt"],
    "response_a": train_texts_ft["response_a"],
    "response_b": train_texts_ft["response_b"],
    "label": y_train_ft
})
val_dataset = Dataset.from_dict({
    "prompt": val_texts["prompt"],
    "response_a": val_texts["response_a"],
    "response_b": val_texts["response_b"],
    "label": y_val_ft
})

# 并行 tokenization，加速
tokenized_train = train_dataset.map(preprocess_function, batched=True, num_proc=2)
tokenized_val = val_dataset.map(preprocess_function, batched=True, num_proc=2)

# 5️⃣ 评估函数
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    preds = logits.argmax(axis=-1)
    return {"accuracy": accuracy_score(labels, preds)}

# 6️⃣ 训练配置（显存优化）
training_args = TrainingArguments(
    output_dir="./ft_results",
    per_device_train_batch_size=8,          # 减小 batch size
    per_device_eval_batch_size=16,
    gradient_accumulation_steps=2,         # 两步累积，相当于 batch_size=16
    num_train_epochs=3,
    learning_rate=3e-4,
    logging_steps=50,
    fp16=True,                              # 半精度训练
    report_to=[]                            # 禁用 W&B
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train,
    eval_dataset=tokenized_val,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

# 7️⃣ 开始微调
print("⏳ 开始 LoRA 微调 ...")
trainer.train()
print("✅ 微调完成。")

# 8️⃣ 测试集预测
test_texts = [
    f"问题: {p} [SEP] A: {a} [SEP] B: {b}" 
    for p, a, b in zip(test["prompt"], test["response_a"], test["response_b"])
]
test_dataset = Dataset.from_dict({"text": test_texts})
tokenized_test = test_dataset.map(lambda x: tokenizer(x["text"], truncation=True, padding="max_length", max_length=256), batched=True, num_proc=2)

pred_logits = trainer.predict(tokenized_test).predictions
pred_probs = torch.softmax(torch.tensor(pred_logits), dim=-1).numpy()

submission_ft = pd.DataFrame(pred_probs, columns=sample.columns[1:])
submission_ft.insert(0, "id", sample["id"])
submission_ft.to_csv("submission_finetuned.csv", index=False)
print("✅ 微调模型结果已保存为 submission_finetuned.csv")



--- 正在执行 Step 3.3: LoRA 微调 (显存优化) ---


/usr/local/lib/python3.11/dist-packages/transformers/convert_slow_tokenizer.py:564: UserWarning: The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option which is not implemented in the fast tokenizers. In practice this means that the fast version of the tokenizer can produce unknown tokens whereas the sentencepiece version would have converted these unknown tokens into a sequence of byte tokens matching the original piece of text.
  warnings.warn(
Some weights of DebertaV2ForSequenceClassification were not initialized from the model checkpoint at ./deberta-small-local and are newly initialized: ['classifier.bias', 'classifier.weight', 'pooler.dense.bias', 'pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map (num_proc=2):   0%|          | 0/55477 [00:00<?, ? examples/s]

Map (num_proc=2):   0%|          | 0/2000 [00:00<?, ? examples/s]

/tmp/ipykernel_37/4068557808.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(
No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


⏳ 开始 LoRA 微调 ...


✅ 微调完成。


Map (num_proc=2):   0%|          | 0/3 [00:00<?, ? examples/s]

✅ 微调模型结果已保存为 submission_finetuned.csv


#### （2）Temperature Scaling 概率校准  
- **原理：** 对 logits 除以温度 T (>0)，以最小化 validation log loss 确定最优 T。  
- **实现：** 使用 `scipy.optimize.minimize` 在 [0.5, 5.0] 范围搜索最优温度。  
- **结果：** 最佳 T ≈ 1.55，校准后模型输出概率分布更加平滑、符合真实置信度。  
- **输出：** 生成 `submission_calibrated.csv` 作为最终校准结果。

In [10]:
# === Calibration (Temperature Scaling) ===
print("\n--- 正在执行: 概率校准 (Temperature Scaling) ---")

from sklearn.metrics import log_loss
from scipy.optimize import minimize
import numpy as np

# 使用验证集 logits
val_logits = trainer.predict(tokenized_val).predictions
val_probs = torch.softmax(torch.tensor(val_logits), dim=-1).numpy()

def temperature_scale(logits, T):
    logits_T = logits / T
    exp_T = np.exp(logits_T - np.max(logits_T, axis=1, keepdims=True))
    return exp_T / np.sum(exp_T, axis=1, keepdims=True)

def loss_fn(T):
    probs_T = temperature_scale(val_logits, T)
    return log_loss(y_val_ft, probs_T)

# 优化温度参数 T
res = minimize(loss_fn, x0=[1.0], bounds=[(0.5, 5.0)], method="L-BFGS-B")
T_opt = res.x[0]
print(f"📏 最优温度参数 T = {T_opt:.3f}")

# 应用到测试集预测
calibrated_probs = temperature_scale(pred_logits, T_opt)

submission_calibrated = pd.DataFrame(calibrated_probs, columns=sample.columns[1:])
submission_calibrated.insert(0, "id", sample["id"])
submission_calibrated.to_csv("submission_calibrated.csv", index=False)
submission_calibrated.to_csv("submission.csv", index=False)
print("✅ 校准后结果已保存为 submission_calibrated.csv")


--- 正在执行: 概率校准 (Temperature Scaling) ---


📏 最优温度参数 T = 1.629
✅ 校准后结果已保存为 submission_calibrated.csv


### ✅ 总结
| 模型类型 | 方法 | 验证 LogLoss |
|-----------|------|---------------|
| MiniLM Embedding + LR | 基线嵌入模型 | 1.0849 |
| E5 + LightGBM | 单模型扩展 | 1.0838 |
| MiniLM + E5 Ensemble | 模型融合 | 1.0767 |
| Bias-aware LGBM + PCA | 加入偏置特征 | 1.0326 |
| DeBERTa-small + LoRA | 轻量级微调 | ≈ 1.07 |
| LoRA + Temperature Scaling | 微调后校准 | ≈ 1.05 (最终提交) |

> 经过模型扩展、偏置建模、轻量级微调与校准等步骤，模型在验证集 LogLoss 上较基线显著下降，性能与可靠性均得到提升。
